src/coreclr/vm/prestub.cpp

```diff
@@ -1250,6 +1271,64 @@ namespace
         return true;
     }
 
+    bool AreConstraintsEqual(const Instantiation& left, const Instantiation& right)
+    {
+        STANDARD_VM_CONTRACT;
+
+        DWORD argCount = left.GetNumArgs();
+        if (argCount != right.GetNumArgs())
+            return false;
+
+        for (DWORD i = 0; i < argCount; ++i)
+        {
+            TypeHandle tL = left[i];
+            TypeHandle tR = right[i];
+
+            // Check generic variable state are the same.
+            BOOL isGeneric = tL.IsGenericVariable();
+            if (isGeneric != tR.IsGenericVariable())
+                return false;
+
+            // Only generic variables have constraints.
+            if (!isGeneric)
+                continue;
+
+            TypeVarTypeDesc* tsL = tL.AsGenericVariable();
+            TypeVarTypeDesc* tsR = tR.AsGenericVariable();
+
+            //
+            // Verify general constraints
+            //
+            IMDInternalImport* importL = tsL->GetModule()->GetMDImport();
+            IMDInternalImport* importR = tsR->GetModule()->GetMDImport();
+
+            DWORD flagsL;
+            DWORD flagsR;
+            IfFailThrow(importL->GetGenericParamProps(tsL->GetToken(), NULL, &flagsL, NULL, NULL, NULL));
+            IfFailThrow(importR->GetGenericParamProps(tsR->GetToken(), NULL, &flagsR, NULL, NULL, NULL));
+            if ((flagsL & gpSpecialConstraintMask) != (flagsR & gpSpecialConstraintMask))
+                return false;
+
+            //
+            // Verify type constraints
+            //
+            DWORD constraintCountL;
+            DWORD constraintCountR;
+            TypeHandle* cL = tsL->GetConstraints(&constraintCountL);
+            TypeHandle* cR = tsR->GetConstraints(&constraintCountR);
+            if (constraintCountL != constraintCountR)
+                return false;
+
+            for (DWORD j = 0; j < constraintCountL; ++j)
+            {
+                if (cL[j] != cR[j])
```

Comment:
The simple comparison does not work for cases like:
```csharp
using System;
using System.Runtime.CompilerServices;

class Accessor
{
    public static void Main() => CallM<string, object>(new C());

    // Actual behavior: "Generic method constraints do not match." exception
    // Expected behavior: Succeeds
    [UnsafeAccessor(UnsafeAccessorKind.Method, Name = "M")]
    extern static void CallM<T,U>(C c) where T:U, IEquatable<T>;
}

class C
{
    public void M<T,U>() where T:U, IEquatable<T>
    {
        Console.WriteLine(typeof(T).ToString());
        Console.WriteLine(typeof(U).ToString());
    }
}
```

Reply:
Do we _need_ to validate the constraints? From my experiments for #97084, we have a well defined failure mode when the constraints in a call don't match (VerificationException). Could we just leave it to that? The exception would be hit when JIT compiling the unsafe accessor.

(Might need to check if we do similar verification for fields, but if not, it would probably not be hard to add it to getFieldInfo - same as in getCallInfo.)

Reply:
> Do we need to validate the constraints?

For this PR, @davidwrighton and I prefer this approach. Once this work is in, I will start on exploring a more relaxed/nuanced constraint checking but precise match is the current design goal.

Reply:
Changing the constraints order is a compatible change. I think that doing precise match as implemented here is fundamentally broken.

The proper way to do static constraints validation would by providing that the UnsafeAccessor constraints are sufficient to satisfy the target method constraints. Same logic as what's used for regular method calls (look for `.SatisfiesConstraints`).

I think that this logic should be either deleted from this PR, or add TODO comment saying this is throw away broken logic that will be reworked, or implement something that is correct.

Reply:
> Changing the constraints order is a compatible change.

But it isn't compatible as defined here. The expectation is the signatures match precisely. This follows the same logic as Type and Method params must be the same index. My goal here is to have the precise match.

> Same logic as what's used for regular method calls (look for .SatisfiesConstraints).

That is a fair approach, but I was under the impression we were going for stricter match. I can consider that instead though.

Reply:
> The expectation is the signatures match precisely.

Yes, changing signatures is a [breaking change](https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/breaking-change-rules.md) so there is no problem with UnsafeAccessor requiring exact signature match.

Changing constraints order is not a breaking change, so UnsafeAccessor should allow it. (I am not sure whether we have it explicitly written down anywhere that changing constraints order is non-breaking change, but our entire existing system treats it as non-breaking.)